Fix errors from issue 5970

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Implement getting arrays from the SMT2 backend

This was previously unimplemented, which put null expressions into the
trace steps and resulted in the elements of arrays not being printed
when the trace is printed.

TGWDB commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -741,13 +750,19 @@ literalt smt2_convt::convert(const exprt &expr)
   convert_literal(l);
   out << " () Bool ";
-  convert_expr(prepared_expr);
+  if(can_cast_expr<forall_exprt>(prepared_expr))
+  {
+    out << "true";
+  }
+  else
```
Should there also be one for `exists_exprt`?

TGWDB commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -741,13 +750,19 @@ literalt smt2_convt::convert(const exprt &expr)
   exprt prepared_expr = prepare_for_convert_expr(expr);
 
   literalt l(no_boolean_variables, false);
+
```
Maybe don't add newlines?

thomasspriggs committed:
Ignore quantifier errors running `get-value` commands with `z3`

Because the other values can yield a usefull trace.

TGWDB committed:
Track conversion and use identifier instead of expression

This tracks when an expression is converted by storing the
identifier in the SMT output in the defined_expressions
map. Then when asserting this expression, merely assert the
identifier and don't duplicate the expression in the output.

thomasspriggs committed:
Use `set_to` to infer missing boolean identifier values

This fixes some of the cases where `z3` returns an error instead of a
value for some of the values cbmc is attempting to get.

thomasspriggs commented:
The original approach in this PR was fairly unsound. It only produced better results for one specific test case and it introduced new problems. I have now pushed a new approach based on ignoring errors from `z3` and inferring the values we couldn't get based the assertions also present in the SMT2 file. This should produce better results when solving with `z3`, where the quantifiers are inside a call to `cprover_assume`. However we do not have an equivalent fix for where the quantifiers are inside an assert.

TGWDB commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -745,6 +745,16 @@ literalt smt2_convt::convert(const exprt &expr)
+  std::stringstream tmpstr;
+  tmpstr << "|B" << l.var_no() << "|";
+  defined_expressions[expr] = tmpstr.str();
+  // End temporary storage
```
This line of comment no longer correct.

TGWDB commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -4314,7 +4324,28 @@ void smt2_convt::set_to(const exprt &expr, bool value)
+  }
+  out << ")\n";
+  return;
+  // OLD BELOW
```
If this fix stays in then we should remove the old code below.

thomasspriggs replied:
Removed. Why didn't you just remove this to begin with? :satisfied:

thomasspriggs replied:
I have refactored the code and removed this comment.

TGWDB commented on src/solvers/smt2/smt2_conv.h:
```diff
@@ -225,6 +225,7 @@ class smt2_convt : public stack_decision_proceduret
 
   typedef std::map<exprt, irep_idt> defined_expressionst;
   defined_expressionst defined_expressions;
+  std::unordered_map<irep_idt, bool> set_values;
```
Doxygen this?

thomasspriggs replied:
Done.

TGWDB reviewed (approved):
Approving as all changes are only tidying up/improving commentary.

thomasspriggs committed:
Remove dead code

thomasspriggs committed:
Tidy up storage of defined literals

thomasspriggs committed:
Add doxygen of `set_values`
